### PR TITLE
[WIP] Avoid truncation when longer output, ref. #6267

### LINKED ISSUE #6267: Avoid truncation when it makes the output longer
Pytest version 3.10.1, also seen on Ubuntu 19.04:

With `-v`, when the output of a failure is close to the truncation limit, verbosity is *increased*, and the extra words are not useful. In short, this:

```
$ py.test-3 -v | wc
     60     269    3287
$ py.test-3 -vv | wc
     59     272    3292
```
In addition the  truncation message is misleading in the number of truncated lines. With `-v`:

```
E       assert b'<!DOCTYPE H...cation.</p>\n' == b'{}'
E         At index 0 diff: 60 != 123
E         Left contains more items, first extra item: 68
E         Full diff:
E         + b'{}'
E         - (b'<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2 Final//EN">\n<title>500 Inter'
E         -  b'nal Server Error</title>\n<h1>Internal Server Error</h1>\n<p>The server en'
E         -  b'countered an internal error and was unable to complete your request.  Either'...
E         
E         ...Full output truncated (2 lines hidden), use '-vv' to show
```

With `-vv`:

```
E       assert b'<!DOCTYPE H...cation.</p>\n' == b'{}'
E         At index 0 diff: 60 != 123
E         Left contains more items, first extra item: 68
E         Full diff:
E         + b'{}'
E         - (b'<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2 Final//EN">\n<title>500 Inter'
E         -  b'nal Server Error</title>\n<h1>Internal Server Error</h1>\n<p>The server en'
E         -  b'countered an internal error and was unable to complete your request.  Either'
E         -  b' the server is overloaded or there is an error in the application.</p>\n')
```

That looks *one* extra line to me, not two. Possibly lines are counted by splitting on `\n`, and the message has a terminating `\n`, technically leaving an empty line at the end -- but this empty line is not shown in the `-vv` output.

it would be better for pytest to print the one line of error, rather than the two lines making an inaccurate claim about the truncation.

### PR DESCRIPTION
**Pull request by erheron**
Fixes #6267 
Surely, because it's user-facing change, documentation/tests would need to change too, but now it's rather a proposal to mentioned issue.

### CODE REVIEW

**Comment by blueyed** [comment]
As for a test I've noticed a change with https://github.com/pytest-dev/pytest/pull/6702/files#diff-14ad89cac128f1dd3aaa927c656035b3L992-R1025 in that regard.
(I was a bit surprised for the explanation to also count towards the lines, but it makes a bit of sense - but is likely more due to there being no concept of "header notes" and "diff output/details")

**Comment by bluetech** [comment]
Hi @erheron,

We've recently stopped using the `features` branch, and use the `master` branch instead ([rationale](https://github.com/pytest-dev/pytest/pull/6571)). We're asking open PRs against `features` to be rabased on `master`. You can do it as follows:

- `git checkout master && git pull origin` - sync the latest code.
- `git checkout avoid-truncation-when-longer-output-6267` - checkout the branch of this PR.
- `git rebase -i master` - leave only your own commits, and delete everything else.
- `git push --force` - update the PR.
- Click `Edit` in the PR GitHub UI and change the base branch from `features` to `master`.

Let us know if you have any issues.

Thanks!

File: src/_pytest/assertion/truncate.py [hunk 1]
```diff
@@ -48,6 +50,7 @@ def _truncate_explanation(input_lines, max_lines=None, max_chars=None):
         max_chars = DEFAULT_MAX_CHARS
 
     # Check if truncation required
+
```
Comment (blueyed) [inline src/_pytest/assertion/truncate.py]:
```suggestion
```

File: src/_pytest/assertion/truncate.py [hunk 2]
```diff
@@ -62,12 +65,18 @@ def _truncate_explanation(input_lines, max_lines=None, max_chars=None):
...
+    last_truncated = not (
+        input_lines[len(truncated_explanation) - 1] in truncated_explanation[-1]
+    )
     if truncated_line_count == 1:
-        msg += " ({} line hidden)".format(truncated_line_count)
+        msg += " ({} line hidden".format(truncated_line_count)
+    else:
+        msg += " ({} lines hidden".format(truncated_line_count)
+    if last_truncated:
+        msg += ",1 truncated)"
```
Comment (blueyed) [inline src/_pytest/assertion/truncate.py]: Looks like a missing space (after comma)?

**Comment by bluetech** [comment]
> We've recently stopped using the features branch

(I've changed the base on my own now).

File: src/_pytest/assertion/truncate.py (hunk 2 above)
Comment (erheron) [reply src/_pytest/assertion/truncate.py]: Yes, the space was missing. I've commited 2 changes, haven't squashed yet

**Comment by nicoddemus** [comment]
Hi @erheron,

First of all we would like to thank you for your time and effort on working on this, the pytest team deeply appreciates it.

We noticed it has been awhile since you have updated this PR, however. pytest is a high activity project, with many issues/PRs being opened daily, so it is hard for us maintainers to track which PRs are ready for merging, for review, or need more attention.

So for those reasons we think it is best to close the PR for now, but with the only intention to cleanup our queue, it is by no means a rejection of your changes. We still encourage you to re-open this PR (it is just a click of a button away) when you are ready to get back to it.

Again we appreciate your time for working on this, and hope you might get back to this at a later time!

Cheers!

**Comment by erheron** [comment]
Hi @nicoddemus. While working on this issue I got a feeling, that this is kind of minor change, and I'm not sure I want to re-open this. However, if someone else decides to work on this issue - please, feel free to use my code, if it would help You in any way!

**Comment by nicoddemus** [comment]
Sure @erheron! No worries, thanks a lot anyway, we appreciate it. 👍

**Pull request closed without merging**